src/otx/algo/object_detection_3d/monodetr3d.py

```diff
@@ -0,0 +1,258 @@
+        # transformer
+        depthaware_transformer = DepthAwareTransformerBuilder(self.model_name)
+        # depth prediction module
+        depth_predictor = DepthPredictor(depth_num_bins=80, depth_min=1e-3, depth_max=60.0, hidden_dim=256)
+        # criterion
+        loss_weight_dict = {
+            "loss_ce": 2,
+            "loss_bbox": 5,
+            "loss_giou": 2,
+            "loss_center": 10,
+        }
+        criterion = MonoDETRCriterion(num_classes=num_classes, focal_alpha=0.25, weight_dict=loss_weight_dict)
+
+        return MonoDETR(
+            backbone,
+            depthaware_transformer,
+            depth_predictor,
+            num_classes=num_classes,
+            criterion=criterion,
+            num_queries=50,
+            aux_loss=True,
+            num_feature_levels=4,
+            with_box_refine=True,
+            init_box=False,
+        )
+
+    def _customize_inputs(
+        self,
+        entity: Det3DBatchDataEntity,
+    ) -> dict[str, Any]:
+        # prepare bboxes for the model
+        targets_list = []
+        img_sizes = torch.from_numpy(np.array([img_info.ori_shape for img_info in entity.imgs_info])).to(
+            device=entity.images.device,
+        )
+        key_list = ["labels", "boxes", "depth", "size_3d", "heading_angle", "boxes_3d"]
+        for bz in range(len(entity.imgs_info)):
+            target_dict = {}
+            for key in key_list:
+                target_dict[key] = getattr(entity, key)[bz]
+            targets_list.append(target_dict)
+
+        return {
+            "images": entity.images,
+            "calibs": torch.cat([p2.unsqueeze(0) for p2 in entity.calib_matrix], dim=0),
+            "targets": targets_list,
+            "img_sizes": img_sizes,
+            "mode": "loss" if self.training else "predict",
+        }
+
+    def _customize_outputs(
+        self,
+        outputs: dict[str, torch.Tensor],
+        inputs: Det3DBatchDataEntity,
+    ) -> Det3DBatchPredEntity | OTXBatchLossEntity:
+        if self.training:
+            if not isinstance(outputs, dict):
+                raise TypeError(outputs)
+
+            losses = OTXBatchLossEntity()
+            for k, v in outputs.items():
+                if isinstance(v, list):
+                    losses[k] = sum(v)
+                elif isinstance(v, Tensor):
+                    losses[k] = v
+                else:
+                    msg = "Loss output should be list or torch.tensor but got {type(v)}"
+                    raise TypeError(msg)
+            return losses
+
+        labels, scores, size_3d, heading_angle, boxes_3d, depth = self.extract_dets_from_outputs(outputs)
+        # bbox 2d decoding
+        boxes_2d = box_cxcylrtb_to_xyxy(boxes_3d)
+        xywh_2d = box_convert(boxes_2d, "xyxy", "cxcywh")
+        # size 2d decoding
+        size_2d = xywh_2d[:, :, 2:4]
+
+        return Det3DBatchPredEntity(
+            batch_size=len(outputs),
```

Comment:
That's incorrect: len(outputs) just counts the number of keys in the dict coming from the model, not batch size

Reply:
done